Tune test_storage_rabbitmq flaky test

pamarcos opened the pull request:
Tune test_storage_rabbitmq to try to fix https://github.com/ClickHouse/ClickHouse/issues/71049:

- Increase max memory usage for RabbitMQ from 2GB to 4GB
- Split the problematic tests into a separate test that runs by itself without any other test in parallel
- Fix `test_attach_broken_table` and `test_rabbitmq_nack_failed_insert` to be able to run them multiple times. The latter now properly restores the original configuration

Close https://github.com/ClickHouse/ClickHouse/issues/71049

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory and will run independently of the checks below:
- [ ] <!---ci_include_stateless--> Only: Stateless tests
- [ ] <!---ci_include_integration--> Only: Integration tests
- [ ] <!---ci_include_performance--> Only: Performance tests
---
- [ ] <!---ci_exclude_style--> Skip: Style check
- [ ] <!---ci_exclude_fast--> Skip: Fast test
---
- [ ] <!---woolen_wolfdog--> Run all checks ignoring all possible failures (Resource-intensive. All test jobs execute in parallel).
- [ ] <!---no_ci_cache--> Disable CI cache

pamarcos committed:
Remove unnecessary call to revive_rabbitmq

All callers of kill_rabbitmq are already calling revive_rabbitmq,
so it's redundant to call that directly (without any sleep).

pamarcos committed:
Increase num_batches from 4->6 and 6->8 for Integration tests

This is an attempt to address the rabbitmq flaky test issue.

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #75656 from ClickHouse/ClickHouse:fix-rabbitmq-flaky-test -->
*This is an automated comment for commit d8e7b45380586bfa8af370518ea8d88d051c92d7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/75656/d8e7b45380586bfa8af370518ea8d88d051c92d7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/73b3b3ad24de7868adf2a2795edad1aa05ffdb7a/builds/report.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

azat reviewed (changes requested):
I guess you are trying to fix the following (test got killed) - https://github.com/ClickHouse/ClickHouse/issues/71049#issuecomment-2634742359

But it is unclear how tests groups works, so it is only a matter of time when it will fail again

How about adding this test into `tests/integration/parallel_skip.json`?

And also I'm pretty sure that rabbitmq can be tuned somehow to make it less memory greedy (if it the RabbitMQ after all), and it is actually already configured to use not more then 2GiB, and with total 64GiB for ClickHouse, it should be OK (of course it is unpredictable with parallel tests, though I'm pretty sure that other ClickHouse servers can it ~2GiB, but nothing is killed)

So TL;DR; let's forbid parallel run for this test

qoega commented:
If we block connectivity via partition manager it is usually good to run not in parallel. We just do not need to have too many non parallel tests

pamarcos commented:
I don't know where the culprit of the issue is, but [memory does not seem to be the one to blame](https://github.com/ClickHouse/ClickHouse/issues/71049#issuecomment-2642281144).

> But it is unclear how tests groups works, so it is only a matter of time when it will fail again

Yep, I'm not sure how batches are created, but AFAIK they're not deterministically split. I also agree increasing the number of batches does not deterministically guarantee the tests will run with more resources, so I think it's worth a try testing to run these tests by themselves. Thanks for the tip.

>and it is actually already configured to use not more then 2GiB

Where do you see that 2GiB limit?

pamarcos committed:
Revert "Increase num_batches from 4->6 and 6->8 for Integration tests"

This reverts commit d8e7b45380586bfa8af370518ea8d88d051c92d7.

pamarcos committed:
Increase max memory usage for RabbitMQ

pamarcos committed:
Add test_storage_rabbitmq to the list of sequential tests

pamarcos committed:
Merge branch 'master' into fix-rabbitmq-flaky-test

pamarcos committed:
Split problematic tests to their own separate test

Run them sequentially to avoid other parallel tests messing
with them.

pamarcos committed:
Merge branch 'master' into fix-rabbitmq-flaky-test

maxknv committed:
Merge branch 'master' into fix-rabbitmq-flaky-test

pamarcos committed:
Merge branch 'master' into fix-rabbitmq-flaky-test

pamarcos committed:
Remove non-existing pytest config: session-timeout

pamarcos committed:
Add back session_timeout with proper name

pamarcos committed:
Merge branch 'master' into fix-rabbitmq-flaky-test

pamarcos committed:
Add split test into parallel_skip.yaml

pamarcos committed:
Use thread method for timeout

It's more reliable according to the documentation:
https://pypi.org/project/pytest-timeout/

at the expense of presenting worse stacks. I've tested
it and the stack is reasonable to pinpoint where it got
stuck.

pamarcos committed:
Minor improvements to test

pamarcos committed:
Use rabbitmq_cluster.wait_rabbitmq_to_start instead of implementing our own

This one calls rabbitmq_debuginfo, which provides extra information
about the status of RabbitMQ's docker instance.

pamarcos committed:
Fix rabbitmq tests so they can run multiple times

pamarcos committed:
Restore original configuration after executing the test

pamarcos commented:
The CI found a LOGICAL_ERROR and also 3 flaky tests that I've created issues for. What's concerning the rabbitmq tests is green. I'm not 100% sure it's going to fix them for good, but it's definitely much better than it was. Please @azat take another look whenever you can.

azat commented on tests/integration/pytest.ini:
```diff
@@ -1,7 +1,9 @@
 norecursedirs = _instances*
+session_timeout = 3300
 timeout = 900
+timeout_method = thread
```
BTW can you confirm that this works?

pamarcos replied:
Yes, of course. I set a low timeout and run the test to confirm:

```bash
❯ tests/integration/runner --cleanup-containers test_storage_rabbitmq_failed_connection/test.py -- -x
2025-02-20 11:27:25,472 [ 147816 ] INFO : ClickHouse root is not set. Will use /home/ubuntu/ClickHouse/ClickHouse (runner:53, check_args_and_update_paths)
2025-02-20 11:27:25,472 [ 147816 ] INFO : Cases dir is not set. Will use /home/ubuntu/ClickHouse/ClickHouse/tests/integration (runner:97, check_args_and_update_paths)
2025-02-20 11:27:25,472 [ 147816 ] INFO : utils dir is not set. Will use /home/ubuntu/ClickHouse/ClickHouse/utils (runner:108, check_args_and_update_paths)
2025-02-20 11:27:25,472 [ 147816 ] INFO : base_configs_dir: /home/ubuntu/ClickHouse/ClickHouse/programs/server/, binary: /home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse, cases_dir: /home/ubuntu/ClickHouse/ClickHouse/tests/integration  (runner:110, check_args_and_update_paths)
clickhouse_integration_tests_volume
Running pytest container as: 'docker run --net=host -it --rm --name clickhouse_integration_tests_37rqfu --privileged --dns-search='.' --memory=63823327232 --volume=/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse-odbc-bridge:/clickhouse-odbc-bridge --volume=/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse:/clickhouse --volume=/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse-library-bridge:/clickhouse-library-bridge --volume=/home/ubuntu/ClickHouse/ClickHouse/programs/server/:/clickhouse-config --volume=/home/ubuntu/ClickHouse/ClickHouse/tests/integration:/ClickHouse/tests/integration --volume=/home/ubuntu/ClickHouse/ClickHouse/utils/backupview:/ClickHouse/utils/backupview --volume=/home/ubuntu/ClickHouse/ClickHouse/utils/grpc-client/pb2:/ClickHouse/utils/grpc-client/pb2 --volume=/run:/run/host:ro --volume=clickhouse_integration_tests_volume:/var/lib/docker  -e PYTEST_CLEANUP_CONTAINERS=1 -e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600  -e PYTHONUNBUFFERED=1 -e PYTEST_ADDOPTS="  test_storage_rabbitmq_failed_connection/test.py -x   -vvv" clickhouse/integration-tests-runner:latest '.
Start tests
========================================================================================== test session starts ==========================================================================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /ClickHouse/tests/integration
configfile: pytest.ini
plugins: reportlog-0.4.0, xdist-3.5.0, random-0.2, repeat-0.9.3, order-1.0.0, timeout-2.3.1
timeout: 5.0s
timeout method: thread
timeout func_only: False
collected 2 items

test_storage_rabbitmq_failed_connection/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Captured stdout ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copy common default production configuration from /clickhouse-config. Files: config.xml, users.xml
Copy common default production configuration from /clickhouse-config. Files: config.xml, users.xml
Copy common default production configuration from /clickhouse-config. Files: config.xml, users.xml
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Captured stderr ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
iptables: Index of deletion too big.
ip6tables: Index of deletion too big.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Stack of MainThread (131614435889152) ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  File "/usr/local/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 192, in console_main
    code = main()
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 169, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 318, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 271, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 325, in _main
    config.hook.pytest_runtestloop(session=session)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 350, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 114, in pytest_runtest_protocol
    runtestprotocol(item, nextitem=nextitem)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 127, in runtestprotocol
    rep = call_and_report(item, "setup", log)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 222, in call_and_report
    call = call_runtest_hook(item, when, **kwds)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 261, in call_runtest_hook
    return CallInfo.from_call(
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 341, in from_call
    result: Optional[TResult] = func()
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 157, in pytest_runtest_setup
    item.session._setupstate.setup(item)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 494, in setup
    col.setup()
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 1795, in setup
    self._request._fillfixtures()
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 566, in _fillfixtures
    item.funcargs[argname] = self.getfixturevalue(argname)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 585, in getfixturevalue
    fixturedef = self._get_active_fixturedef(argname)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 607, in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 693, in _compute_fixture_value
    fixturedef.execute(request=subrequest)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 1069, in execute
    result = ihook.pytest_fixture_setup(fixturedef=self, request=request)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 1123, in pytest_fixture_setup
    result = call_fixture_func(fixturefunc, request, kwargs)
  File "/usr/local/lib/python3.10/dist-packages/_pytest/fixtures.py", line 895, in call_fixture_func
    fixture_result = next(generator)
  File "/ClickHouse/tests/integration/test_storage_rabbitmq_failed_connection/test.py", line 81, in rabbitmq_cluster
    cluster.start()
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 3065, in start
    run_and_check(clickhouse_start_cmd)
  File "/ClickHouse/tests/integration/helpers/cluster.py", line 131, in run_and_check
    res = subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 505, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1154, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.10/subprocess.py", line 2021, in _communicate
    ready = selector.select(timeout)
  File "/usr/lib/python3.10/selectors.py", line 416, in select
    fd_event_list = self._selector.poll(timeout)
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

azat replied:
Great, as for the corner cases we will test it in master :)

Can you please also add a comment in pytest.ini, so that if someone will want to change it, he/she will have to exaplin it additionally

pamarcos replied:
Done in [`da96286` (#75656)](https://github.com/ClickHouse/ClickHouse/pull/75656/commits/da96286f8f3ef66d0c2beb80bc6ae367d5c82695)

pamarcos commented on tests/integration/test_storage_rabbitmq_failed_connection/configs/users.xml:
```diff
@@ -0,0 +1,14 @@
+<clickhouse>
```
Sure. That's what I'd have done if I had seen any other integration test with two test files. I didn't see any, so I mimicked what we had. I'll add only the `test_failed_connection.py` and set `parallel_skip.yaml` so that all `test_storage_rabbitmq` tests run sequentially then.

azat replied:
>I didn't see any,

Just in case, here is I good example:

```
$ ls -1 tests/integration/test_backward_compatibility/test_*.py
tests/integration/test_backward_compatibility/test_aggregate_fixed_key.py
tests/integration/test_backward_compatibility/test_aggregate_function_state.py
tests/integration/test_backward_compatibility/test_convert_ordinary.py
tests/integration/test_backward_compatibility/test_cte_distributed.py
tests/integration/test_backward_compatibility/test_functions.py
tests/integration/test_backward_compatibility/test_insert_profile_events.py
tests/integration/test_backward_compatibility/test_ip_types_binary_compatibility.py
tests/integration/test_backward_compatibility/test_memory_bound_aggregation.py
tests/integration/test_backward_compatibility/test_normalized_count_comparison.py
tests/integration/test_backward_compatibility/test_parallel_replicas_protocol.py
tests/integration/test_backward_compatibility/test_select_aggregate_alias_column.py
tests/integration/test_backward_compatibility/test_short_strings_aggregation.py
tests/integration/test_backward_compatibility/test_vertical_merges_from_compact_parts.py
```

pamarcos replied:
Done in [`da96286` (#75656)](https://github.com/ClickHouse/ClickHouse/pull/75656/commits/da96286f8f3ef66d0c2beb80bc6ae367d5c82695)

pamarcos committed:
Move test_failed_connection.py within test_storage_rabbitmq dir

azat reviewed (approved):
LGTM

Though I would update the description, to something like (since it is unclear will it fix the test or not) - `Tune test_storage_rabbitmq`

azat commented:
>[Integration tests (asan, old analyzer, 1/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=75656&sha=da96286f8f3ef66d0c2beb80bc6ae367d5c82695&name_0=PR&name_1=Integration%20tests%20%28asan%2C%20old%20analyzer%2C%201%2F6%29) — Job timeout expired, fail: 100, passed: 443

Interesting, why it is `Job timeout`, even though the i.e. `test_storage_azure_blob_storage/test_cluster.py::test_union_all` (for which there are no results in the report) completed - https://s3.amazonaws.com/clickhouse-test-reports/PRs/75656/da96286f8f3ef66d0c2beb80bc6ae367d5c82695//integration_tests_asan_old_analyzer_1_6/integration_run_parallel4_0.log

azat commented:
>@pamarcos
[Add pytest-fork to ensure JUnit XML output completes](https://github.com/ClickHouse/ClickHouse/pull/75656/commits/f9dd2714b07fd1f67e4d37787ab4835c621cb5b3)

:+1:

pamarcos committed:
Restore signal method for timeout

pamarcos commented:
I've reverted the changes I did to use thread as timeout_method because getting that to work was not as quick as I though. I'll address it in a separate PR so that we can merge this one ASAP to get rid of the disturbing flaky tests
